pkg/identity/identitysync/azure.go

```diff
@@ -0,0 +1,254 @@
+	OdataNextLink *string      `json:"@odata.nextLink,omitempty"`
+	Value         []AzureGroup `json:"value"`
+}
+
+type AzureGroup struct {
+	ID          string `json:"id"`
+	Description string `json:"description"`
+	DisplayName string `json:"displayName"`
+}
+
+type UserGroups struct {
+	OdataNextLink *string  `json:"@odata.nextLink,omitempty"`
+	OdataContext  string   `json:"@odata.context"`
+	Value         []string `json:"value"`
+}
+
+type ClientSecretCredential struct {
+	client confidential.Client
+}
+
+// GetToken requests an access token from Azure Active Directory. This method is called automatically by Azure SDK clients.
+func (c *ClientSecretCredential) GetToken(ctx context.Context) (string, error) {
+	ar, err := c.client.AcquireTokenByCredential(ctx, []string{"https://graph.microsoft.com/.default"})
+	return ar.AccessToken, err
+}
+
+func NewClientSecretCredential(s deploy.Azure, httpClient *http.Client) (*ClientSecretCredential, error) {
+	cred, err := confidential.NewCredFromSecret(s.ClientSecret)
+	if err != nil {
+		return nil, err
+	}
+	c, err := confidential.New(s.ClientID, cred,
+		confidential.WithAuthority(fmt.Sprintf("%s/%s", ADAuthorityHost, s.TenantID)),
+		confidential.WithHTTPClient(httpClient))
+	if err != nil {
+		return nil, err
+	}
+	return &ClientSecretCredential{client: c}, nil
+}
+
+// NewAzure will fail if the Azure settings are not configured
+func NewAzure(ctx context.Context, settings deploy.Azure) (*AzureSync, error) {
+	azAuth, err := NewClientSecretCredential(settings, http.DefaultClient)
+	if err != nil {
+		return nil, err
+	}
+	token, err := azAuth.GetToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AzureSync{NewClient: http.DefaultClient, token: token}, nil
+}
+
+func (a *AzureSync) Get(url string) (*http.Response, error) {
+	req, _ := http.NewRequest("GET", url, nil)
+	req.Header.Add("Authorization", "Bearer "+a.token)
+	return a.NewClient.Do(req)
+
+}
+
+// idpUserFromAzureUser converts a azure user to the identityprovider interface user type
+func (a *AzureSync) idpUserFromAzureUser(ctx context.Context, azureUser AzureUser) (identity.IdpUser, error) {
+	u := identity.IdpUser{
+		ID:        azureUser.ID,
+		FirstName: azureUser.GivenName,
+		LastName:  azureUser.Surname,
+		Email:     azureUser.Mail,
+		Groups:    []string{},
+	}
+
+	g, err := a.GetMemberGroups(u.ID)
+	if err != nil {
+		return identity.IdpUser{}, err
+	}
+	u.Groups = g
+
+	return u, nil
+}
+
+func (a *AzureSync) GetMemberGroups(userID string) ([]string, error) {
+	var userGroups []string
+
+	hasMore := true
+	var nextToken *string
+	url := MSGraphBaseURL + fmt.Sprintf("/directoryObjects/%s/getMemberGroups", userID)
+
+	for hasMore {
+		var jsonStr = []byte(`{ "securityEnabledOnly": false}`)
+		req, _ := http.NewRequest("POST", url, bytes.NewBuffer(jsonStr))
+		req.Header.Add("Authorization", "Bearer "+a.token)
+		req.Header.Set("Content-Type", "application/json")
+
+		res, err := a.NewClient.Do(req)
```

Comment:
I think that you will need to add a status check e.g res.StatusCode == 200
else return an error.
This goes for all your requests